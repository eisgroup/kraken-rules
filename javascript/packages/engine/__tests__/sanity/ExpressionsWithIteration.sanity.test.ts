/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { sanityEngine } from "./_SanityEngine";
import { sanityMocks } from "./_AutoPolicyObject.mocks";

describe("Engine Expressions Sanity Test", () => {
    const { empty } = sanityMocks;
    it("shouldEvaluateDeeplyNestedLoopExpression", () => {
        const policy = empty();

        const results = sanityEngine.evaluate(policy, "Expressions_nested_for");

        expect(results).k_toHaveExpressionsFailures(0);
        expect(policy.transactionDetails!.totalPremium).toBe(123);
        expect(results).k_toMatchResultsSnapshots();
    });
    it("shouldEvaluateDeeplyNestedFilterExpression", () => {
        const policy = empty();
        policy.riskItems = [
            {
                id: "1",
                cd: "Vehicle",
                odometerReading : 100000
            }
        ];

        const results = sanityEngine.evaluate(policy, "Expressions_nested_filter");

        expect(results).k_toHaveExpressionsFailures(0);
        expect(policy.termDetails!.termNo).toBe(100000);
        expect(results).k_toMatchResultsSnapshots();
    });
    it("shouldEvaluateDeeplyNestedMixedExpression", () => {
        const policy = empty();
        policy.riskItems = [
            {
                id: "1",
                cd: "Vehicle",
                odometerReading : 100000
            }
        ];

        const results = sanityEngine.evaluate(policy, "Expressions_nested_mixed");

        expect(results).k_toHaveExpressionsFailures(0);
        expect(policy.transactionDetails!.changePremium).toBe(100000);
        expect(results).k_toMatchResultsSnapshots();
    });
});
