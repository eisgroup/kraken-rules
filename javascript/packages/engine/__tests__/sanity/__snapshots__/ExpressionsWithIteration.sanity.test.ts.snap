// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engine Expressions Sanity Test shouldEvaluateDeeplyNestedFilterExpression 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:0:termNo": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "termNo",
          "fieldPath": "termDetails.termNo",
          "fieldType": "INTEGER",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "termDetails.termNo",
                  "contextId": "0",
                  "contextName": "Policy",
                  "newValue": 100000,
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "DEFAULT",
              "ruleName": "Expressions_nested_filter",
              "targetPath": "termNo",
            },
          },
        ],
      },
    },
    "ruleTimeZoneId": "Europe/Vilnius",
  },
  "fieldResults": Object {
    "Policy:0:termNo": Object {
      "fieldType": "INTEGER",
      "id": "Policy:0:termNo",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "termDetails.termNo",
      "ruleResults": Array [],
    },
  },
}
`;

exports[`Engine Expressions Sanity Test shouldEvaluateDeeplyNestedLoopExpression 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:0:totalPremium": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "totalPremium",
          "fieldPath": "transactionDetails.totalPremium",
          "fieldType": "DECIMAL",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "transactionDetails.totalPremium",
                  "contextId": "0",
                  "contextName": "Policy",
                  "newValue": 123,
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "DEFAULT",
              "ruleName": "Expressions_nested_for",
              "targetPath": "totalPremium",
            },
          },
        ],
      },
    },
    "ruleTimeZoneId": "Europe/Vilnius",
  },
  "fieldResults": Object {
    "Policy:0:totalPremium": Object {
      "fieldType": "DECIMAL",
      "id": "Policy:0:totalPremium",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "transactionDetails.totalPremium",
      "ruleResults": Array [],
    },
  },
}
`;

exports[`Engine Expressions Sanity Test shouldEvaluateDeeplyNestedMixedExpression 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:0:changePremium": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "changePremium",
          "fieldPath": "transactionDetails.changePremium",
          "fieldType": "DECIMAL",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "transactionDetails.changePremium",
                  "contextId": "0",
                  "contextName": "Policy",
                  "newValue": 100000,
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "DEFAULT",
              "ruleName": "Expressions_nested_mixed",
              "targetPath": "changePremium",
            },
          },
        ],
      },
    },
    "ruleTimeZoneId": "Europe/Vilnius",
  },
  "fieldResults": Object {
    "Policy:0:changePremium": Object {
      "fieldType": "DECIMAL",
      "id": "Policy:0:changePremium",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "transactionDetails.changePremium",
      "ruleResults": Array [],
    },
  },
}
`;
